### PR TITLE
[network/ebpf/http] http_parse_data() : using __builtin_memcmp()/__builtin_strlen()

### PR DESCRIPTION
**Pull request by nplanel**
Compare tests are in this branch
https://github.com/DataDog/datadog-agent/tree/nplanel/ebpf-fast-compare-less-instructions

The conclusion is the __builtin_strcmp() use less instructions

Results :

-DU64
    1915:	05 00 f4 ff 00 00 00 00	goto -12 <LBB0_155>
-DUX
    1913:	05 00 f4 ff 00 00 00 00	goto -12 <LBB0_155>
-DU64U32
    1926:	05 00 f4 ff 00 00 00 00	goto -12 <LBB0_156>
-DBUILTIN
    1804:	05 00 ef ff 00 00 00 00	goto -17 <LBB0_156>
-DBUILTINSTRLEN
    1804:	05 00 ef ff 00 00 00 00	goto -17 <LBB0_156>
-DBUILTINSTRCMP
    1793:	05 00 f4 ff 00 00 00 00	goto -12 <LBB0_156>
-DORIG
    1884:	05 00 35 ff 00 00 00 00	goto -203 <LBB0_94>

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by p-lambert — COMMENTED** [review]
This is great, but I recall having some issues when trying to use `__builtin_memcmp`.
At least there used to be a bug in the LLVM eBPF backend. [See Use of LLVM built-in functions](https://docs.cilium.io/en/v1.8/bpf/)
Right now it looks like that your program is getting rejected by the verifier:
> program socket__http_filter: can't unmarshal instructions: offset 2184: can't relocate instruction: load: .L.str.1: unsupported binding: STB_LOCAL

**Comment by brycekahle** [comment]
We are not on the latest version of Clang, so if the inlining is the problem, we could try upgrading Clang first.

File: pkg/network/ebpf/c/http.h [hunk 1]
```diff
@@ -141,30 +141,32 @@ static __always_inline int http_begin_response(http_transaction_t *http, const c
 }
 
 static __always_inline void http_parse_data(char *p, http_packet_t *packet_type, http_method_t *method) {
-    if ((p[0] == 'H') && (p[1] == 'T') && (p[2] == 'T') && (p[3] == 'P')) {
+#define __builtin_const_strcmp(buf, constbuf) __builtin_memcmp(buf, constbuf, __builtin_strlen(constbuf))
```
Comment (brycekahle) [inline pkg/network/ebpf/c/http.h:144]: Have you examined the instructions to see if the compiler is smart enough to replace `__builtin_strlen` with a constant value?

Comment (nplanel) [reply pkg/network/ebpf/c/http.h:144]: is the tests, `__builtin_strlen()` vs `constant value` generate the same numbers of instructions

**Comment by nplanel** [comment]
> Details

yeah I just seen it this morning that llvm reference the static strings in a section, but latest (IIRC) llvm/clang can handle this case for us.

**Comment by nplanel** [comment]
> We are not on the latest version of Clang, so if the inlining is the problem, we could try upgrading Clang first.

yeah IIRC latest clang handle this correctly in ebpf (not creating a section for static const strings)

**Comment by brycekahle** [comment]
> latest (IIRC) llvm/clang can handle this case for us.

Can you test this in your agent VM?

**Comment by nplanel** [comment]
> 

Yep sure I will

**Pull request closed without merging**